### PR TITLE
Implement basic 64-bit float operations to amd64 compiler backend.

### PR DESCRIPTION
**Pull request by twitchyliquid64**
This PR implements a bunch more operations around floats to the amd64 backend. I have not run any benchmarks as we dont have any benches that exercise floating-point operations.

It should be safe to use SSE2 instructions as I am doing, as they will be present on all CPUs that Go supports.

The handling of NaN when performing comparison operations is a bit #yolo, let me know if you have any ideas that dont involve branching.

### CODE REVIEW

**Comment by sbinet** [comment]
Just a quick note: I am off the grid (-ish). Will be back on Tuesday.
(But sounds interesting!)

**Comment by twitchyliquid64** [comment]
No worries at all :) Enjoy going dark ;)

File: exec/internal/compile/backend_amd64.go [hunk 1]
```diff
@@ -544,6 +578,127 @@ func (b *AMD64Backend) emitBinaryI64(builder *asm.Builder, regs *dirtyRegs, ci c
...
+	}
+	writeNaNVal.To.Type = obj.TYPE_REG
+	writeNaNVal.To.Reg = x86.REG_AX
+	writeNaNVal.As = x86.AMOVQ
+	jmpNaN.Pcond = writeNaNVal
+	builder.AddInstruction(writeNaNVal)
+
+	// Symbolic instruction so the not-NaN case can avoid being
+	// overwritten. Normal flow (!NaN) results in a jump to here.
+	// The assembler will optimize this puesdo-instruction so as
```
[the remark this answers is not in the document]
Comment (twitchyliquid64) [reply exec/internal/compile/backend_amd64.go]: Oops! I always do that, along with `recieve`.

File: exec/internal/compile/backend_amd64_test.go [hunk 2]
```diff
@@ -916,6 +917,426 @@ func TestAMD64OperationsI32(t *testing.T) {
...
+			Result: math.Float64bits(-2),
+		},
+		{
+			Name:   "sub",
+			Op:     ops.F64Sub,
+			Args:   []uint64{math.Float64bits(12), math.Float64bits(3)},
+			Result: math.Float64bits(9),
+		},
+		{
+			Name:   "sub-negative",
```
[the remark this answers is not in the document]
Comment (twitchyliquid64) [reply exec/internal/compile/backend_amd64_test.go]: Added.

File: exec/internal/compile/backend_amd64_test.go [hunk 3]
```diff
@@ -916,6 +917,426 @@ func TestAMD64OperationsI32(t *testing.T) {
 	}
 }
 
+func TestAMD64OperationsF64(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (twitchyliquid64) [reply exec/internal/compile/backend_amd64_test.go]: Hmmm, just looked this up, its messy.

So for the normal binary ops returning a float, the result of an op with a NaN operand is non-deterministic. In practice, I expect that whatever the SSE2 execution unit does is probably what everyone else does, given everyone is JIT'ing anyway.

For the comparison ops, the output when NaNs are involved is [defined](https://webassembly.github.io/spec/core/_download/WebAssembly.pdf) starting at page 61. TL;DR: Return true for NaN for not-equal, and false for all other kinds of comparison. There are tests for these cases, and the spec tests in `exec/` also cover these as well.

[the remark this answers is not in the document]
Comment (twitchyliquid64) [reply exec/internal/compile/backend_amd64_test.go]: Should we check for / handle any NaN behaviours outside of comparison ops, or just assume we leave it at NaN propagation like the SSE2 engine does?

[the remark this answers is not in the document]
Comment (twitchyliquid64) [reply exec/internal/compile/backend_amd64_test.go]: Double checked our implementation & various language semantics - from what I can tell, most specifications (and our Go implementation) rely on NaN propergation for binary ops, and only specify behaviour in the case of NaN on comparison ops & a few special functions like Min. I believe this is consistent with this PR.

**Pull request merged**